splice_machine/src/main/java/com/splicemachine/derby/stream/function/QuoteTrackingTokenizer.java

```diff
@@ -69,211 +140,309 @@
+     */
+    public QuoteTrackingTokenizer(final Reader reader, final CsvPreference preferences, final List<Integer> valueSizeHints) {
+        super(reader, preferences);
+        this.quoteChar = preferences.getQuoteChar();
+        this.delimeterChar = preferences.getDelimiterChar();
+        this.surroundingSpacesNeedQuotes = preferences.isSurroundingSpacesNeedQuotes();
+        this.ignoreEmptyLines = preferences.isIgnoreEmptyLines();
+        this.commentMatcher = preferences.getCommentMatcher();
+        this.maxLinesPerRow = preferences.getMaxLinesPerRow();
+        this.valueSizeHints = valueSizeHints;
     }
 
     @Override
-    public boolean readColumns(final List<String> columns) throws IOException{
-        return readColumns(columns,null);
+    public boolean readColumns(final List<String> columns) throws IOException {
+        return readColumns(columns, null);
     }
 
-    public boolean readColumns(final List<String> columns,final BooleanList quotedColumns) throws IOException{
-        if(columns==null){
+    private void reset() {
+        // clear the reusable List and StringBuilders
+        colIdx = 0;
+        currentColumn.setLength(0);
+        // process each character in the line, catering for surrounding quotes (QUOTE_MODE)
+        state = TokenizerState.NORMAL;
+        quoteScopeStartingLine = -1;
+        potentialSpaces = 0;
+        charIndex = 0;
+        wasQuoted = false;
+    }
+
+    /**
+     * Parses the columns from current CSV row.
+     *
+     * @param columns       output parameter, the resulting set of parsed columns
+     * @param quotedColumns output parameter, for each one of the resulting columns: true if the column was quoted, otherwise, false
+     * @return true if more rows exists, otherwise false.
+     */
+    public boolean readColumns(final List<String> columns, final BooleanList quotedColumns) throws IOException {
+        if (columns == null) {
             throw new NullPointerException("columns should not be null");
         }
-        boolean recordQuotes = quotedColumns!=null;
-
-        // clear the reusable List and StringBuilders
         columns.clear();
-        if(recordQuotes) quotedColumns.clear();
-        currentColumn.setLength(0);
-        currentRow.setLength(0);
+        if (quotedColumns != null) quotedColumns.clear();
+
+        reset();
+        currentRow.clear();
 
         // read a line (ignoring empty lines/comments if necessary)
-        String line;
-        do{
-            line=readLine();
-            if(line==null){
-                return false; // EOF
+        if (!readFirstLine()) return false; // EOF
+
+        // now, if some column value is larger than scanThreshold we perform the dry run to avoid extra allocation of StringBuilder.
+        checkExactSizeFirst = false;
+        if (valueSizeHints != null) {
+            if (valueSizeHints.stream().max(Integer::compare).get() > scanThresold) {
```

Comment:
>> Could we add another check here? If we are parsing with oneLineRecords we could check the length of the current line, if it's less than the scanThreshold we avoid the dryRun too

> I think at this point of execution we don't know the actual size of the current CSV line, or maybe I missed some info in your comment?

We can get the size of the current line with `getCurrentLine().length()` don't we?
What we don't know is whether we are parsing in `oneLineRecords` mode, we'd need to pass down that flag into the QuoteTrackingTokenizer, or have a helper method in MutableCSVTokenizer (that one only gets used in `oneLineRecords` mode)

Reply:
done, it made the parsing logic slightly more complicated but it is totally worth it. Thanks!